File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/offline/OfflineBalanceProof.java
package org.hyperledger.iroha.android.offline;

/** Native helpers for computing offline balance commitments and proofs. */
public final class OfflineBalanceProof {
  private static final boolean NATIVE_AVAILABLE;

  static {
    boolean available;
    try {
      System.loadLibrary("connect_norito_bridge");
      available = true;
    } catch (UnsatisfiedLinkError error) {
      available = false;
    }
    NATIVE_AVAILABLE = available;
  }

  private OfflineBalanceProof() {}

  public static boolean isNativeAvailable() {
    return NATIVE_AVAILABLE;
  }

  public static byte[] updateCommitment(
      final byte[] initialCommitment,
      final String claimedDelta,
      final byte[] initialBlinding,
      final byte[] resultingBlinding) {
    if (!NATIVE_AVAILABLE) {
      throw new IllegalStateException("connect_norito_bridge is not available in this runtime");
    }
    return nativeUpdateCommitment(initialCommitment, claimedDelta, initialBlinding, resultingBlinding);
  }

  public static byte[] generateProof(
      final String chainId,
      final byte[] initialCommitment,
      final byte[] resultingCommitment,
      final String claimedDelta,
      final String resultingValue,
      final byte[] initialBlinding,
      final byte[] resultingBlinding) {
    if (!NATIVE_AVAILABLE) {
      throw new IllegalStateException("connect_norito_bridge is not available in this runtime");
    }
    return nativeGenerate(
        chainId,
        initialCommitment,
        resultingCommitment,
        claimedDelta,
        resultingValue,
        initialBlinding,
        resultingBlinding);
  }

  public static Artifacts advanceCommitment(
      final String chainId,
      final String claimedDelta,
      final String resultingValue,
      final String initialCommitmentHex,
      final String initialBlindingHex,
      final String resultingBlindingHex) {
    final byte[] initialCommitment = decodeHex(initialCommitmentHex, "initialCommitmentHex");
    final byte[] initialBlinding = decodeHex(initialBlindingHex, "initialBlindingHex");
    final byte[] resultingBlinding = decodeHex(resultingBlindingHex, "resultingBlindingHex");
    final byte[] resultingCommitment =
        updateCommitment(initialCommitment, claimedDelta, initialBlinding, resultingBlinding);
    final byte[] proof =
        generateProof(
            chainId,
            initialCommitment,
            resultingCommitment,
            claimedDelta,
            resultingValue,
            initialBlinding,
            resultingBlinding);
    return new Artifacts(resultingCommitment, proof);
  }

  public static final class Artifacts {
    private final byte[] resultingCommitment;
    private final byte[] proof;

    private Artifacts(final byte[] resultingCommitment, final byte[] proof) {
      this.resultingCommitment = resultingCommitment;
      this.proof = proof;
    }

    public byte[] getResultingCommitment() {
      return resultingCommitment.clone();
    }

    public byte[] getProof() {
      return proof.clone();
    }

    public String getResultingCommitmentHex() {
      return encodeHex(resultingCommitment);
    }

    public String getProofHex() {
      return encodeHex(proof);
    }
  }

  private static native byte[] nativeUpdateCommitment(
      byte[] initialCommitment,
      String claimedDelta,
      byte[] initialBlinding,
      byte[] resultingBlinding);

  private static native byte[] nativeGenerate(
      String chainId,
      byte[] initialCommitment,
      byte[] resultingCommitment,
      String claimedDelta,
      String resultingValue,
      byte[] initialBlinding,
      byte[] resultingBlinding);

  private static byte[] decodeHex(final String hex, final String field) {
    final String trimmed = hex == null ? "" : hex.trim();
    if ((trimmed.length() & 1) == 1) {
      throw new IllegalArgumentException(field + " must contain an even number of hex characters");
    }
    final byte[] out = new byte[trimmed.length() / 2];
    for (int i = 0; i < trimmed.length(); i += 2) {
      final int hi = Character.digit(trimmed.charAt(i), 16);
      final int lo = Character.digit(trimmed.charAt(i + 1), 16);
      if (hi == -1 || lo == -1) {
        throw new IllegalArgumentException(field + " contains non-hex characters");
      }
      out[i / 2] = (byte) ((hi << 4) + lo);
    }
    return out;
  }

  private static String encodeHex(final byte[] bytes) {
    final StringBuilder builder = new StringBuilder(bytes.length * 2);
    for (final byte b : bytes) {
      builder.append(String.format("%02x", b));
    }
    return builder.toString();
  }
}

File: kotlin/offline-wallet-android/src/main/java/org/hyperledger/iroha/sdk/offline/OfflineBalanceProof.kt
package org.hyperledger.iroha.sdk.offline

/** Native helpers for computing offline balance commitments and proofs. */
class OfflineBalanceProof private constructor() {

    class Artifacts internal constructor(
        resultingCommitment: ByteArray,
        proof: ByteArray,
    ) {
        private val _resultingCommitment: ByteArray = resultingCommitment.copyOf()
        private val _proof: ByteArray = proof.copyOf()

        val resultingCommitment: ByteArray get() = _resultingCommitment.copyOf()
        val proof: ByteArray get() = _proof.copyOf()

        val resultingCommitmentHex: String get() = encodeHex(_resultingCommitment)
        val proofHex: String get() = encodeHex(_proof)
    }

    companion object {
        private val NATIVE_AVAILABLE: Boolean

        init {
            var available: Boolean
            try {
                System.loadLibrary("connect_norito_bridge")
                available = true
            } catch (_: UnsatisfiedLinkError) {
                available = false
            }
            NATIVE_AVAILABLE = available
        }

        @JvmStatic
        fun isNativeAvailable(): Boolean = NATIVE_AVAILABLE

        @JvmStatic
        fun updateCommitment(
            initialCommitment: ByteArray,
            claimedDelta: String,
            initialBlinding: ByteArray,
            resultingBlinding: ByteArray,
        ): ByteArray {
            check(NATIVE_AVAILABLE) { "connect_norito_bridge is not available in this runtime" }
            return nativeUpdateCommitment(initialCommitment, claimedDelta, initialBlinding, resultingBlinding)
        }

        @JvmStatic
        fun generateProof(
            chainId: String,
            initialCommitment: ByteArray,
            resultingCommitment: ByteArray,
            claimedDelta: String,
            resultingValue: String,
            initialBlinding: ByteArray,
            resultingBlinding: ByteArray,
        ): ByteArray {
            check(NATIVE_AVAILABLE) { "connect_norito_bridge is not available in this runtime" }
            return nativeGenerate(
                chainId,
                initialCommitment,
                resultingCommitment,
                claimedDelta,
                resultingValue,
                initialBlinding,
                resultingBlinding,
            )
        }

        @JvmStatic
        fun advanceCommitment(
            chainId: String,
            claimedDelta: String,
            resultingValue: String,
            initialCommitmentHex: String,
            initialBlindingHex: String,
            resultingBlindingHex: String,
        ): Artifacts {
            val initialCommitment = decodeHex(initialCommitmentHex, "initialCommitmentHex")
            val initialBlinding = decodeHex(initialBlindingHex, "initialBlindingHex")
            val resultingBlinding = decodeHex(resultingBlindingHex, "resultingBlindingHex")
            val resultingCommitment =
                updateCommitment(initialCommitment, claimedDelta, initialBlinding, resultingBlinding)
            val proof = generateProof(
                chainId,
                initialCommitment,
                resultingCommitment,
                claimedDelta,
                resultingValue,
                initialBlinding,
                resultingBlinding,
            )
            return Artifacts(resultingCommitment, proof)
        }

        @JvmStatic
        private external fun nativeUpdateCommitment(
            initialCommitment: ByteArray,
            claimedDelta: String,
            initialBlinding: ByteArray,
            resultingBlinding: ByteArray,
        ): ByteArray

        @JvmStatic
        private external fun nativeGenerate(
            chainId: String,
            initialCommitment: ByteArray,
            resultingCommitment: ByteArray,
            claimedDelta: String,
            resultingValue: String,
            initialBlinding: ByteArray,
            resultingBlinding: ByteArray,
        ): ByteArray

        private fun decodeHex(hex: String?, field: String): ByteArray {
            val trimmed = hex?.trim() ?: ""
            require(trimmed.length % 2 == 0) { "$field must contain an even number of hex characters" }
            val out = ByteArray(trimmed.length / 2)
            for (i in trimmed.indices step 2) {
                val hi = Character.digit(trimmed[i], 16)
                val lo = Character.digit(trimmed[i + 1], 16)
                require(hi != -1 && lo != -1) { "$field contains non-hex characters" }
                out[i / 2] = ((hi shl 4) + lo).toByte()
            }
            return out
        }

        private fun encodeHex(bytes: ByteArray): String {
            val builder = StringBuilder(bytes.size * 2)
            for (b in bytes) {
                builder.append(String.format("%02x", b))
            }
            return builder.toString()
        }
    }
}
